README.md
Waveform
========

Waveform is a class to generate waveform images from audio files. You can combine it with jPlayer to make a soundcloud.com style MP3 player. It also comes with a handy CLI you can use to generate waveform images on the command line.

Installation
============

Waveform depends on `ruby-audio`, which in turn depends on libsndfile.

Build libsndfile from (http://www.mega-nerd.com/libsndfile/), install it via `apt` (`sudo apt-get install libsndfile1-dev`), `libsndfile` in macports, etc...

Then:

    $ sudo gem install waveform

Image creation depends on `chunky_png`, which has a faster native library called `oily_png` which will be used if available.

    $ sudo gem install oily_png

NOTE: If `ruby-audio` fails to compile and you have `libsndfile` available, it may be because of this: http://stackoverflow.com/questions/19919640/ruby-audio-1-6-1-install-error-with-installed-libsndfile-1-0-25

CLI Usage
=========

    $ waveform song.wav waveform.png

There are some nifty options you can supply to switch things up:

    -W sets the width (in pixels) of the waveform image.
    -H sets the height (in pixels).
    -c sets the color used to draw the waveform (in hex, can also use
        'transparent').
    -b sets the background color to draw the waveform on (in hex, and can use
        'transparent' as well).
    -m sets the method used to sample the source audio file, it can either be
        'peak' or 'rms'. 'peak' is probably what you want because it looks
        cooler, but 'rms' is closer to what you actually hear.

There are also some less-nifty options:

    -q will generate your waveform without printing out a bunch of stuff.
    -h will print out a help screen with all this info.
    -F will automatically overwrite destination file.

Generating a small waveform "cut out" of a white background is pretty useful,
then you can overlay it on a web-gradient on the website for your new startup
and it will look really cool. To make it you could use:

    $ waveform -W900 -H140 -ctransparent -b#ffffff Motley\ Crüe/Kickstart\ my\ Heart.wav sweet_waveforms/Kickstart\ my\ Heart.png

Usage in code
=============

The CLI is really just a thin wrapper around the Waveform class, which you can also use in your programs for reasons I haven't thought of. The Waveform class takes pretty much the same options as the CLI when generating waveforms.

Requirements
============

`ruby-audio`

The gem version, *not* the old outdated library listed on RAA. `ruby-audio` is a wrapper for `libsndfile`, on my Ubuntu 10.04LTS VM I installed the necessary libs to build `ruby-audio` via: `sudo apt-get install libsndfile1-dev`.

`chunky_png`

`chunky_png` is a pure ruby (!) PNG manipulation library. Caveat to this requirement is that if you also install `oily_png` you will get *better performance* as it uses some C code, and C code is fast.

Converting MP3 to WAV
=====================

Waveform used to (very thinly) wrap ffmpeg to convert MP3 (and whatever other format) to WAV audio before processing the WAV and generating the waveform image. It seemed a bit presumptious for Waveform to handle that, especially since you might want to use your own conversion options (i.e. downsampling the bitrate to generate waveforms faster, etc...).

If you happen to be using ffmpeg, you can easily convert MP3 to WAV via:

    ffmpeg -i "/path/to/source/file.mp3" -f wav "/path/to/output/file.wav"

Tests
=====

    $ rake

If you get an error about not being able to find ruby-audio gem (and you have ruby-audio gem) you might need to let rake know how to load your gems -- if you're using rubygems:

    $ export RUBYOPT="rubygems"
    $ rake

Sample sound file used in tests is in the Public Domain from soundbible.com: <http://soundbible.com/1598-Electronic-Chime.html>.

Changes
=======

0.1.3
-----
  * Update for ruby 3.2 (replace `File.exists?` w/ `File.exist?`)

0.1.2
-----
  * Added more helpful deprecation notice for non-WAV audio files

0.1.1
-----
  * Fixed RMS calculation (was calculating RMSD instead of RMS before) -- thanks, cviedmai

0.1.0
-----
  * No more wrapping ffmpeg to automatically convert mp3 to wav
  * Fixed support for mono audio sources (4-channel, surround, etc. should also work)
  * Change to gemspec & added seperate version file so that bundler won't try to load ruby-audio (thanks, amiel)
  * Changed Waveform-class API as there's no longer a need to instantiate a waveform

References
==========

<http://pscode.org/javadoc/src-html/org/pscode/ui/audiotrace/AudioPlotPanel.html#line.996>
<http://github.com/pangdudu/rude/blob/master/lib/waveform_narray_testing.rb>
<http://stackoverflow.com/questions/1931952/asp-net-create-waveform-image-from-mp3>
<http://codeidol.com/java/swing/Audio/Build-an-Audio-Waveform-Display>

License
=======

Copyright (c) 2010-2012 Ben Alavi

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
## 0.1.3 (2023-03-22)

- ruby 3.2 compatability (replace File.exists? w/ File.exist?)


test/waveform_test.rb
require File.expand_path(File.join(File.dirname(__FILE__), "..", "lib", "waveform"))

require "test/unit"
require "fileutils"

module Helpers
  def fixture_asset(file)
    File.join(File.dirname(__FILE__), "fixtures", file)
  end

  def output(file)
    File.join(File.dirname(__FILE__), "output", file)
  end

  def open_png(file)
    ChunkyPNG::Image.from_datastream(ChunkyPNG::Datastream.from_file(file))
  end
end

class WaveformTest < Test::Unit::TestCase
  include Helpers
  extend Helpers

  def self.cleanup
    puts "Removing existing testing artifacts..."
    Dir[output("*.*")].each{ |f| FileUtils.rm(f) }
    FileUtils.mkdir_p(output(""))
  end

  def test_generates_waveform
    Waveform.generate(fixture_asset("sample.wav"), output("waveform_from_audio_source.png"))
    assert File.exist?(output("waveform_from_audio_source.png"))

    image = open_png(output("waveform_from_audio_source.png"))
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:color]), image[60, 120]
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:background_color]), image[0, 0]
  end

  def test_generates_waveform_from_mono_audio_source_via_peak
    Waveform.generate(fixture_asset("mono_sample.wav"), output("waveform_from_mono_audio_source_via_peak.png"))
    assert File.exist?(output("waveform_from_mono_audio_source_via_peak.png"))

    image = open_png(output("waveform_from_mono_audio_source_via_peak.png"))
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:color]), image[60, 120]
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:background_color]), image[0, 0]
  end

  def test_generates_waveform_from_mono_audio_source_via_rms
    Waveform.generate(fixture_asset("mono_sample.wav"), output("waveform_from_mono_audio_source_via_rms.png"), :method => :rms)
    assert File.exist?(output("waveform_from_mono_audio_source_via_rms.png"))

    image = open_png(output("waveform_from_mono_audio_source_via_rms.png"))
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:color]), image[60, 120]
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:background_color]), image[0, 0]
  end

  def test_logs_to_given_io
    File.open(output("waveform.log"), "w") do |io|
      Waveform.generate(fixture_asset("sample.wav"), output("logged.png"), :logger => io)
    end

    assert_match /Generated waveform/, File.read(output("waveform.log"))
  end

  def test_uses_rms_instead_of_peak
    Waveform.generate(fixture_asset("sample.wav"), output("peak.png"))
    Waveform.generate(fixture_asset("sample.wav"), output("rms.png"), :method => :rms)

    rms = open_png(output("rms.png"))
    peak = open_png(output("peak.png"))

    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:color]), peak[44, 43]
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:background_color]), rms[44, 43]
    assert_equal ChunkyPNG::Color.from_hex(Waveform::DefaultOptions[:color]), rms[60, 120]
  end

  def test_is_900px_wide
    Waveform.generate(fixture_asset("sample.wav"), output("width-900.png"), :width => 900)

    image = open_png(output("width-900.png"))

    assert_equal 900, image.width
  end

  def test_is_100px_tall
    Waveform.generate(fixture_asset("sample.wav"), output("height-100.png"), :height => 100)

    image = open_png(output("height-100.png"))

    assert_equal 100, image.height
  end

  def test_has_auto_width
    Waveform.generate(fixture_asset("sample.wav"), output("width-auto.png"), :auto_width => 10)

    image = open_png(output("width-auto.png"))

    assert_equal 209, image.width
  end

  def test_has_red_background_color
    Waveform.generate(fixture_asset("sample.wav"), output("background_color-#ff0000.png"), :background_color => "#ff0000")

    image = open_png(output("background_color-#ff0000.png"))

    assert_equal ChunkyPNG::Color.from_hex("#ff0000"), image[0, 0]
  end

  def test_has_transparent_background_color
    Waveform.generate(fixture_asset("sample.wav"), output("background_color-transparent.png"), :background_color => :transparent)

    image = open_png(output("background_color-transparent.png"))

    assert_equal ChunkyPNG::Color::TRANSPARENT, image[0, 0]
  end

  def test_has_black_foreground_color
    Waveform.generate(fixture_asset("sample.wav"), output("color-#000000.png"), :color => "#000000")

    image = open_png(output("color-#000000.png"))

    assert_equal ChunkyPNG::Color.from_hex("#000000"), image[60, 120]
  end

  def test_has_red_background_color_with_transparent_foreground_cutout
    Waveform.generate(fixture_asset("sample.wav"), output("background_color-#ff0000+color-transparent.png"), :background_color => "#ff0000", :color => :transparent)

    image = open_png(output("background_color-#ff0000+color-transparent.png"))

    assert_equal ChunkyPNG::Color.from_hex("#ff0000"), image[0, 0]
    assert_equal ChunkyPNG::Color::TRANSPARENT, image[60, 120]
  end

  # Bright green is our transparency mask color, so this test ensures that we
  # don't destroy the image if the background also uses the transparency mask
  # color
  def test_has_transparent_foreground_on_bright_green_background
    Waveform.generate(fixture_asset("sample.wav"), output("background_color-#00ff00+color-transparent.png"), :background_color => "#00ff00", :color => :transparent)

    image = open_png(output("background_color-#00ff00+color-transparent.png"))

    assert_equal ChunkyPNG::Color.from_hex("#00ff00"), image[0, 0]
    assert_equal ChunkyPNG::Color::TRANSPARENT, image[60, 120]
  end

  def test_raises_error_if_not_given_readable_audio_source
    assert_raise(Waveform::RuntimeError) do
      Waveform.generate(fixture_asset("sample.txt"), output("shouldnt_exist.png"))
    end
  end

  def test_overwrites_existing_waveform_if_force_is_true_and_file_exists
    FileUtils.touch output("overwritten.png")

    Waveform.generate(fixture_asset("sample.wav"), output("overwritten.png"), :force => true)
  end

  def test_raises_exception_if_waveform_exists_and_force_is_false
    FileUtils.touch output("wont_be_overwritten.png")

    assert_raises Waveform::RuntimeError do
      Waveform.generate(fixture_asset("sample.wav"), output("wont_be_overwritten.png"), :force => false)
    end
  end

  def test_raises_exception_if_waveform_exists
    FileUtils.touch output("wont_be_overwritten_by_default.png")

    assert_raises Waveform::RuntimeError do
      Waveform.generate(fixture_asset("sample.wav"), output("wont_be_overwritten_by_default.png"))
    end
  end

  def test_raises_deprecation_exception_if_ruby_audio_fails_to_read_source_file
    begin
      Waveform.generate(fixture_asset("sample.txt"), output("shouldnt_exist.png"))
    rescue Waveform::RuntimeError => e
      assert_match /Hint: non-WAV files are no longer supported, convert to WAV first using something like ffmpeg/, e.message
    end
  end
end

WaveformTest.cleanup


waveformer.rb
require "rubygems"
require "ruby-audio"
require "fileutils"
require "rmagick"

# TODO: set these with command line options
# pixel width of final waveform image
Width  = 1800
# pixel height of final waveform image
Height = 280

# A simple wrapper for logging + benchmarking, nice to have good feedback on a
# long batch operation
class Log
  class << self
    # Prints the given message to the log
    def out(msg)
      STDOUT.print(msg)
    end
    
    # Prints the given message to the log followed by the most recent benchmark
    # (note that it calls .end! which will stop the benchmark)
    def done!(msg="done")
      STDOUT.print("#{msg} (#{self.end!}s)\n")
    end
    
    # Starts a new benchmark clock and returns the index of the new clock.
    # 
    # If .start! is called again before .end! then the time returned will be
    # the elapsed time from the next call to start!, and calling .end! again
    # will return the time from *this* call to start! (that is, the clocks are
    # LIFO)
    def start!
      (@benchmarks ||= []) << Time.now
      @current = @benchmarks.size - 1
    end
    
    # Returns the elapsed time from the most recently started benchmark clock
    # and ends the benchmark, so that a subsequent call to .end! will return
    # the elapsed time from the previously started benchmark clock.
    def end!
      elapsed = (Time.now - @benchmarks[@current])
      @current -= 1
      elapsed
    end
    
    # Returns the elapsed time from the benchmark clock w/ the given index (as
    # returned from when .start! was called).
    def time?(index)
      Time.now - @benchmarks[index]
    end
  end
end

class Wave
  def initialize(wave)
    @wave = wave
  end
  
  # Returns a sampling of frames from the given wave file using the given method
  # the sample size is determined by the given pixel width -- we want one sample
  # frame per horizontal pixel.
  def frames(width, method = :peak)
    Log.start!
    Log.out("Analyzing waveform:\n")
  
    frames = []
  
    RubyAudio::Sound.open(@wave) do |snd|
      frames_read       = 0
      frames_per_sample = (snd.info.frames.to_f / Width.to_f).to_i
      sample            = RubyAudio::Buffer.new("float", frames_per_sample, snd.info.channels)

      Log.out("Sampling #{frames_per_sample} frames per sample:\n")
      while(frames_read = snd.read(sample)) > 0
        frames << send(method, sample, snd.info.channels)
        Log.out(".")      
      end
    end

    Log.out("\nRead #{frames.size} frames\n")
    Log.done!
  
    frames
  end
  
  private
  
  # Returns an array of the peak of each channel for the given collection of
  # frames -- the peak is individual to the channel, and the returned collection
  # of peaks are not (necessarily) from the same frame(s).
  def peak(frames, channels=1)
    peak_frame = []
    (0..channels-1).each do |channel|
      peak_frame << channel_peak(frames, channel)
    end
    peak_frame
  end

  # Returns an array of rms values for the given frameset where each rms value is
  # the rms value for that channel.
  def rms(frames, channels=1)
    rms_frame = []
    (0..channels-1).each do |channel|
      rms_frame << channel_rms(frames, channel)
    end
    rms_frame
  end
  
  # Returns the peak voltage reached on the given channel in the given collection
  # of frames.
  # 
  # TODO: Could lose some resolution and only sample every other frame, would
  # likely still generate the same waveform as the waveform is so comparitively
  # low resolution to the original input (in most cases), and would double the
  # analyzation speed
  def channel_peak(frames, channel=0)
    peak = 0.0
    frames.each do |frame|
      next if frame.nil?
      peak = frame[channel] if frame[channel] > peak
    end
    peak
  end

  # Returns the rms value across the given collection of frames for the given
  # channel.
  # 
  # FIXME: this RMS calculation might be wrong...
  # refactored this from: http://pscode.org/javadoc/src-html/org/pscode/ui/audiotrace/AudioPlotPanel.html#line.996
  def channel_rms(frames, channel=0)
    avg = frames.inject(0.0){ |sum, frame| sum += frame ? frame[channel] : 0 }/frames.size.to_f
    Math.sqrt(frames.inject(0.0){ |sum, frame| sum += frame ? (frame[channel]-avg)**2 : 0 }/frames.size.to_f)
  end
end

# Draws a waveform for the given set of sample frames to the given filename
# supports a 2-dimensional array of samples representing multiple channels of
# audio.
def draw(samples, filename)
  Log.start!
  Log.out("Drawing waveform graph:\n")
  
  canvas = Magick::Image.new(Width, Height) { self.background_color = "#ffffff" }
  gc     = Magick::Draw.new
  
  # TODO: set colors from command line options, support more than 2 colors, add
  # default colors per channel
  colors = %w( 000000 000000 )
  
  Log.out("#{samples.size} samples x #{samples[0].size} channels\n")
  
  samples.each_with_index do |sample, x|
    sample.each_with_index do |value, channel|
      scaled = value*Height.to_f
      bottom = (Height-scaled)/2
      
      gc.stroke("##{colors[channel]}")
      gc.stroke_antialias(false)
      gc.stroke_width(1)
      gc.line(x, bottom, x, bottom+scaled)
    end
    Log.out(".")
  end
  
  Log.out("\nDrawing...")
  gc.draw(canvas)
  Log.out("done\n")
  
  Log.out("Finalizing...")
  canvas.transparent("#000000", Magick::TransparentOpacity).write(filename)
  Log.out("done\n")

  Log.done!
end

# Decode given mp3 to a wav file, returns true if the decode succeeded or false
# otherwise.
def mp3_to_wav(mp3, wav)
  Log.start!
  Log.out("Decoding MP3...")
  
  FileUtils.rm(wav) if File.exists?(wav)
  
  system %Q{ffmpeg -i "#{ARGV[0]}" -f wav "#{wav}" > /dev/null 2>&1}
  Log.done!
  
  File.exists?(wav)
end

# =============================================================================

time = Time.now
wave = File.join(File.dirname(__FILE__), "temp.wav")
Log.start!

raise "Unable to decode source MP3 #{ARGV[0]}, quitting" unless mp3_to_wav(ARGV[0], wave)
draw(Wave.new(wave).frames(Width, :peak), ARGV[1])

puts "Generated #{ARGV[1]} in #{(Log.end!)}s"

lib/waveform.rb
require File.join(File.dirname(__FILE__), "waveform/version")

require "ruby-audio"
begin
  require "oily_png"
rescue LoadError
  require "chunky_png"
end

class Waveform
  DefaultOptions = {
    :method => :peak,
    :width => 1800,
    :height => 280,
    :background_color => "#666666",
    :color => "#00ccff",
    :force => false,
    :logger => nil
  }

  TransparencyMask = "#00ff00"
  TransparencyAlternate = "#ffff00" # in case the mask is the background color!

  attr_reader :source

  # Scope these under Waveform so you can catch the ones generated by just this
  # class.
  class RuntimeError < ::RuntimeError;end;
  class ArgumentError < ::ArgumentError;end;

  class << self
    # Generate a Waveform image at the given filename with the given options.
    #
    # Available options (all optional) are:
    #
    #   :method => The method used to read sample frames, available methods
    #     are peak and rms. peak is probably what you're used to seeing, it uses
    #     the maximum amplitude per sample to generate the waveform, so the
    #     waveform looks more dynamic. RMS gives a more fluid waveform and
    #     probably more accurately reflects what you hear, but isn't as
    #     pronounced (typically).
    #
    #     Can be :rms or :peak
    #     Default is :peak.
    #
    #   :width => The width (in pixels) of the final waveform image.
    #     Default is 1800.
    #
    #   :height => The height (in pixels) of the final waveform image.
    #     Default is 280.
    #
    #   :auto_width => msec per pixel. This will overwrite the width of the
    #     final waveform image depending on the length of the audio file.
    #     Example:
    #       100 => 1 pixel per 100 msec; a one minute audio file will result in a width of 600 pixels
    #
    #   :background_color => Hex code of the background color of the generated
    #     waveform image.
    #     Default is #666666 (gray).
    #
    #   :color => Hex code of the color to draw the waveform, or can pass
    #     :transparent to render the waveform transparent (use w/ a solid
    #     color background to achieve a "cutout" effect).
    #     Default is #00ccff (cyan-ish).
    #
    #   :force => Force generation of waveform, overwriting WAV or PNG file.
    #
    #   :logger => IOStream to log progress to.
    #
    # Example:
    #   Waveform.generate("Kickstart My Heart.wav", "Kickstart My Heart.png")
    #   Waveform.generate("Kickstart My Heart.wav", "Kickstart My Heart.png", :method => :rms)
    #   Waveform.generate("Kickstart My Heart.wav", "Kickstart My Heart.png", :color => "#ff00ff", :logger => $stdout)
    #
    def generate(source, filename, options={})
      options = DefaultOptions.merge(options)

      raise ArgumentError.new("No source audio filename given, must be an existing sound file.") unless source
      raise ArgumentError.new("No destination filename given for waveform") unless filename
      raise RuntimeError.new("Source audio file '#{source}' not found.") unless File.exist?(source)
      raise RuntimeError.new("Destination file #{filename} exists. Use --force if you want to automatically remove it.") if File.exist?(filename) && !options[:force] === true

      @log = Log.new(options[:logger])
      @log.start!

      if options[:auto_width]
        RubyAudio::Sound.open(source) do |audio|
          options[:width] = (audio.info.length * 1000 / options[:auto_width].to_i).ceil
        end
      end

      # Frames gives the amplitudes for each channel, for our waveform we're
      # saying the "visual" amplitude is the average of the amplitude across all
      # the channels. This might be a little weird w/ the "peak" method if the
      # frames are very wide (i.e. the image width is very small) -- I *think*
      # the larger the frames are, the more "peaky" the waveform should get,
      # perhaps to the point of inaccurately reflecting the actual sound.
      samples = frames(source, options[:width], options[:method]).collect do |frame|
        frame.inject(0.0) { |sum, peak| sum + peak } / frame.size
      end

      @log.timed("\nDrawing...") do
        # Don't remove the file even if force is true until we're sure the
        # source was readable
        if File.exist?(filename) && options[:force] === true
          @log.out("Output file #{filename} encountered. Removing.")
          File.unlink(filename)
        end

        image = draw samples, options
        image.save filename
      end

      @log.done!("Generated waveform '#{filename}'")
    end

    private

    # Returns a sampling of frames from the given RubyAudio::Sound using the
    # given method the sample size is determined by the given pixel width --
    # we want one sample frame per horizontal pixel.
    def frames(source, width, method = :peak)
      raise ArgumentError.new("Unknown sampling method #{method}") unless [ :peak, :rms ].include?(method)

      frames = []

      RubyAudio::Sound.open(source) do |audio|
        frames_read = 0
        frames_per_sample = (audio.info.frames.to_f / width.to_f).to_i
        sample = RubyAudio::Buffer.new("float", frames_per_sample, audio.info.channels)

        @log.timed("Sampling #{frames_per_sample} frames per sample: ") do
          while(frames_read = audio.read(sample)) > 0
            frames << send(method, sample, audio.info.channels)
            @log.out(".")
          end
        end
      end

      frames
    rescue RubyAudio::Error => e
      raise e unless e.message == "File contains data in an unknown format."
      raise Waveform::RuntimeError.new("Source audio file #{source} could not be read by RubyAudio library -- Hint: non-WAV files are no longer supported, convert to WAV first using something like ffmpeg (RubyAudio: #{e.message})")
    end

    # Draws the given samples using the given options, returns a ChunkyPNG::Image.
    def draw(samples, options)
      image = ChunkyPNG::Image.new(options[:width], options[:height],
        options[:background_color] == :transparent ? ChunkyPNG::Color::TRANSPARENT : options[:background_color]
      )

      if options[:color] == :transparent
        color = transparent = ChunkyPNG::Color.from_hex(
          # Have to do this little bit because it's possible the color we were
          # intending to use a transparency mask *is* the background color, and
          # then we'd end up wiping out the whole image.
          options[:background_color].downcase == TransparencyMask ? TransparencyAlternate : TransparencyMask
        )
      else
        color = ChunkyPNG::Color.from_hex(options[:color])
      end

      # Calling "zero" the middle of the waveform, like there's positive and
      # negative amplitude
      zero = options[:height] / 2.0

      samples.each_with_index do |sample, x|
        # Half the amplitude goes above zero, half below
        amplitude = sample * options[:height].to_f / 2.0
        # If you give ChunkyPNG floats for pixel positions all sorts of things
        # go haywire.
        image.line(x, (zero - amplitude).round, x, (zero + amplitude).round, color)
      end

      # Simple transparency masking, it just loops over every pixel and makes
      # ones which match the transparency mask color completely clear.
      if transparent
        (0..image.width - 1).each do |x|
          (0..image.height - 1).each do |y|
            image[x, y] = ChunkyPNG::Color.rgba(0, 0, 0, 0) if image[x, y] == transparent
          end
        end
      end

      image
    end

    # Returns an array of the peak of each channel for the given collection of
    # frames -- the peak is individual to the channel, and the returned collection
    # of peaks are not (necessarily) from the same frame(s).
    def peak(frames, channels=1)
      peak_frame = []
      (0..channels-1).each do |channel|
        peak_frame << channel_peak(frames, channel)
      end
      peak_frame
    end

    # Returns an array of rms values for the given frameset where each rms value is
    # the rms value for that channel.
    def rms(frames, channels=1)
      rms_frame = []
      (0..channels-1).each do |channel|
        rms_frame << channel_rms(frames, channel)
      end
      rms_frame
    end

    # Returns the peak voltage reached on the given channel in the given collection
    # of frames.
    #
    # TODO: Could lose some resolution and only sample every other frame, would
    # likely still generate the same waveform as the waveform is so comparitively
    # low resolution to the original input (in most cases), and would increase
    # the analyzation speed (maybe).
    def channel_peak(frames, channel=0)
      peak = 0.0
      frames.each do |frame|
        next if frame.nil?
        frame = Array(frame)
        peak = frame[channel].abs if frame[channel].abs > peak
      end
      peak
    end

    # Returns the rms value across the given collection of frames for the given
    # channel.
    def channel_rms(frames, channel=0)
      Math.sqrt(frames.inject(0.0){ |sum, frame| sum += (frame ? Array(frame)[channel] ** 2 : 0) } / frames.size)
    end
  end
end

class Waveform
  # A simple class for logging + benchmarking, nice to have good feedback on a
  # long batch operation.
  #
  # There's probably 10,000,000 other bechmarking classes, but writing this was
  # easier than using Google.
  class Log
    attr_accessor :io

    def initialize(io=$stdout)
      @io = io
    end

    # Prints the given message to the log
    def out(msg)
      io.print(msg) if io
    end

    # Prints the given message to the log followed by the most recent benchmark
    # (note that it calls .end! which will stop the benchmark)
    def done!(msg="")
      out "#{msg} (#{self.end!}s)\n"
    end

    # Starts a new benchmark clock and returns the index of the new clock.
    #
    # If .start! is called again before .end! then the time returned will be
    # the elapsed time from the next call to start!, and calling .end! again
    # will return the time from *this* call to start! (that is, the clocks are
    # LIFO)
    def start!
      (@benchmarks ||= []) << Time.now
      @current = @benchmarks.size - 1
    end

    # Returns the elapsed time from the most recently started benchmark clock
    # and ends the benchmark, so that a subsequent call to .end! will return
    # the elapsed time from the previously started benchmark clock.
    def end!
      elapsed = (Time.now - @benchmarks[@current])
      @current -= 1
      elapsed
    end

    # Returns the elapsed time from the benchmark clock w/ the given index (as
    # returned from when .start! was called).
    def time?(index)
      Time.now - @benchmarks[index]
    end

    # Benchmarks the given block, printing out the given message first (if
    # given).
    def timed(message=nil, &block)
      start!
      out(message) if message
      yield
      done!
    end
  end
end
